### DISCUSSION TITLE
how to handle multiple base urls with rest-client?

### DISCUSSION THREAD

**Discussion by PyAntony** [quarkusio/quarkus#25146] [Q&A]
Hi,

I have an application where I get multiple urls with different base urls. I don't know the base URLs so I can't  add them all in the config file. What I am currently doing is creating a new client for each base URL. Something like:

```java
String[] baseAndRelativePaths = splitUrl(newUrl)

// cache is a mapping from String (base URL) -> web client
cache.putIfAbsent(baseAndRelativePaths[0], getNewClient())

// WebClient.class is the interface
private WebClient getNewClient(String baseUrl) {
  return RestClientBuilder.newBuilder()
        .baseUri(URI.create(basseUrl))
        .property("io.quarkus.rest.client.connection-pool-size", 50)
        .build(WebClient.class)

cache.get(baseAndRelativePaths[0]).doGETrequest(baseAndRelativePaths[1])
}
```

This is working, but I am not sure if this is inefficient. I guess if there is a new thread pool being created for ach base URL this is very inefficient. So:

1) How are connection pools created? one for each base URL or one single connection pool for all the clients?
2) is my approach correct? can it be optimized?
3) Also, does Netty create a new connection for each host or for each full URL? For example, given these 2 urls with the same host:
  - https://example.com/api/endpoint1
  - https://example.com/api/endpoint2
  
Is one connection created or 2?

[a comment by an automated account is not shown]

**Comment by michalszynkiewicz** [comment]
To save resources, you can share the http client between two rest clients.
To do this, set the [`QuarkusRestClientProperties.SHARED` and `QuarkusRestClientProperties.NAME`](https://github.com/quarkusio/quarkus/blob/main/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/api/QuarkusRestClientProperties.java#L33-L43) properties on the client builder:

```java
RestClientBuilder.newBuilder()
        .baseUri(URI.create(basseUrl))
        .property("io.quarkus.rest.client.connection-pool-size", 50)
        .build(WebClient.class)
        .property(QuarkusRestClientProperties.NAME, "my-client-group")
        .property(QuarkusRestClientProperties.SHARED, true);
```

Under the hood, with this setting, Quarkus will use [shared Vert.x HTTP Client](https://vertx.io/docs/vertx-core/java/#_client_sharing)

**Reply by PyAntony** [reply]
So something like:

```java
String[] baseAndRelativePaths = splitUrl(newUrl)

// create new client each time.
// under the hood it's only 1 single http client
var restClient = RestClientBuilder.newBuilder()
        .baseUri(URI.create(baseAndRelativePaths[0]))
        .property("io.quarkus.rest.client.connection-pool-size", 50)
        .build(WebClient.class)
        .property(QuarkusRestClientProperties.NAME, "my-client-group")
        .property(QuarkusRestClientProperties.SHARED, true);

restClient.doGETRequest(baseAndRelativePaths[1])
```

like that?

**Reply by michalszynkiewicz** [reply]
yes, but I'd still put them in a map/cache as you suggested in the original description (if they repeat and you can make sure it won't create a memory leak).

`.property(...)` has to go before `build()` though

**Reply by PyAntony** [reply]
ok, and do you know if Netty create a new connection for each host or for each full URL? I ask cause I have seen one connection created for each URL (even if they all have the same host). I wonder if I am doing something wrong or maybe that's just the way Netty (or any web client) works?

**Reply by michalszynkiewicz** [reply]
I don't know. 
Even with shared client you see one connection for each URL?

**Reply by PyAntony** [reply]
I don't know. I saw 1 connection per URL weeks ago when working with Spring WebClient. I have to see now. Is there a way to access the Netty client directly by the way? Even when activating `quarkus.rest-client.logging.scope=all` I don't see enough logs. When using Spring WebClient I had the following:
```java
HttpClient httpClient = HttpClient.create(connectionProvider)
            .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, connectionTimeout)
            .responseTimeout(Duration.ofMillis(responseTimeout))
            // for each request an redirect log the entire request class and connection
            .doOnRequest((r, c) -> log.debug("HttpClient - sending HTTP. Request, conn: {}, {}.", r, c))
            .doOnRedirect((r, c) -> log.debug("HttpClient - redirecting request. Response, conn: {}, {}.", r, c))
            .followRedirect(true);
```
That's how I knew which connection is handling which request. Any way to do something similar with reastEasy?

**Reply by michalszynkiewicz** [reply]
Oh, I didn't ask the most important question, do you use `quarkus-rest-client-reactive`?

**Reply by PyAntony** [reply]
Yes, I use the following:
```
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-rest-client-reactive-jackson</artifactId>
    </dependency>
```

**Reply by michalszynkiewicz** [reply]
then `quarkus.rest-client.logging.scope=all` should enable netty logging... Maybe some logging category needs to be set to TRACE/DEBUG...

**Reply by alex-martel** [reply]
I am curious about this as well. I know netty is not the apache httpclient. But I wonder if they have the same recommendation as apache [here](https://hc.apache.org/httpclient-legacy/performance.html);
`Generally it is recommended to have a single instance of HttpClient per communication component or even per application`

I tried just "for fun" bundling all my RestClients under the same `QuarkusRestClientProperties.NAME` and set shared to `true`. I do not see any difference in the netty debug logs, well, as @PyAntony said, not enough information to say if it acutally helps. Via tcpdump I see the same with or without that change.

Also unsure how the connection-pool-size work in that case. My rest clients are all configured via config file, and all have a pool size set, if we share some of them which connection-pool would be taken?

**Reply by michalszynkiewicz** [reply]
@PyAntony can you attach a very simple project for me to try it out? I'd be happy to check it but don't have much time...

**Reply by PyAntony** [reply]
A very simple project of the rest-client? do you need to see the logs?

**Reply by michalszynkiewicz** [reply]
one project, two clients, with similar base uris and paths as you use.

**Comment by PyAntony** [comment]
@michalszynkiewicz 
Project here: https://github.com/PyAntony/quarkus-rest-client-demo
You can just git-clone it and run it with `quarkus dev`.

Notice the `application.properties` file:
```
quarkus.rest-client.logging.scope=all

quarkus.log.min-level=TRACE
quarkus.log.category."org.acme".level=TRACE
quarkus.log.category."org.jboss.resteasy.reactive.client.logging".level=TRACE
```
It only logs info about the request and response, but no low level info about Netty.

All this project does is to send a GET request every 5 seconds:
```java
@Scheduled(
        every = "5s",
        concurrentExecution = SKIP
    )
    public void runClient() {
        var urls = List.of(
            "https://docs.python.org/3/whatsnew/3.10.html",
            "https://docs.python.org/3/whatsnew/3.9.html",
            "https://docs.scala-lang.org/tour/higher-order-functions.html",
            "https://docs.scala-lang.org/tour/traits.html"
        );

        urls.forEach(
            url -> webClient.doGETRequest(url)
                .subscribe().with(resp -> {
                    if (resp != null)
                        Log.debugf("Received %s bytes...", resp.getBytes().length);
                    else
                        Log.debugf("UNI produced null...");
                })
        );
    }
```

**Comment by akil-rails** [comment]
I've the same scenario, was there a resolution to this?
